xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -0,0 +1,128 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+
+define(["jquery"], function ($) {
+
+  /**
+   * Map the element to its data object
+   * So that each instance of the livedata on the page handle there own data
+   */
+  var instancesMap = new WeakMap();
+
+
+
+  /**
+   * The init function of the logic script
+   * For each livedata element on the page, returns its corresponding data / API
+   * If the data does not exists yet, create it from the element
+   * @param {HTMLElement} element The HTML Element corresponding to the Livedata component
+   */
+  var init = function (element) {
+
+    if (!instancesMap.has(element)) {
+      // create a new logic object associated to the element
+      var logic = new Logic(element);
+      instancesMap.set(element, logic);
+
+      logic.changeLayout();
+    }
+
+    return instancesMap.get(element);
+  };
+
+
+  /**
+   * Class for a logic element
+   * Contains the Livedata data object and methods to mutate it
+   * Can be used in the layouts to display the data, and call its API
+   * @param {HTMLElement} element The HTML Element corresponding to the Livedata
+   */
+  var Logic = function (element) {
+    this.element = element;
+    this.data = JSON.parse(element.getAttribute("data-data") || "{}");
+    element.removeAttribute("data-data");
+    this.layouts = {};
+  };
+
+
+  /**
+   * Change layout to specified one, or default one if none specified
+   * @param {String} layoutId The id of the layout to load with requireJS
+   * @returns {Object} A jquery promise
+   */
+  Logic.prototype.changeLayout = function (layoutId) {
+    var self = this;
+    var defer = $.Deferred();
+
+    layoutId = layoutId || this.data.meta.defaultLayout;
+    if (layoutId === this.data.query.currentLayout) { return defer.reject(); }
```

Comment:
Yes, indeed we should be able to run code after the function succeeded (layout changed or stayed the same).
But in the first case the event is not send and I wanted to have the same behavior for the promise and the event. However I didn't want the event to be triggered when the layout didn't changed, because this is not intuitive (for instance, the native `change` event does is not triggered when an input has been modified but the value didn't changed).
So I think I'll change the return to resolve, and keep the event system as it is (triggered only when the layout actually changed)

Reply:
> So I think I'll change the return to resolve, and keep the event system as it is (triggered only when the layout actually changed)

Definitely.